[Flang][MLIR] - Handle the mapping of subroutine arguments when they are subsequently used inside the region of  an `omp.target` Op

bhandarkar-pranav opened the pull request:
This is a fix for https://github.com/llvm/llvm-project/issues/134912 which is a problem with mapping `fir.boxchar<k>` type values to the target i.e an `omp.target` op.

There really are two problems. Fixing the first exposed the second. The first problem is that OpenMP lowering of maps in `omp.target` in Flang cannot handle the mapping of a value that doesnt have a defining operation. In other words, a value that is a block argument. This is a corner case and it happens when a value is used inside the region of an `omp.target` and that value is a block argument in the region enclosing the `omp.target` op.  This patch fixes this by trying its best to not use a block argument inside the omp.target if it is possible.

The second problem this fixes is that it adds bounds to `omp.map.info` ops that map `fir.char<k, ?>` types by extracting the length from the corresponding `fir.boxchar`

bhandarkar-pranav committed:
end-to-end test8 working. Need to get rid of lots of debugging commits

bhandarkar-pranav committed:
Removed unconditional debugging prints

bhandarkar-pranav committed:
More cleanup

bhandarkar-pranav committed:
add a testcase

bhandarkar-pranav committed:
clean up som more

bhandarkar-pranav committed:
clean up flang/lib/Lower/OpenMP/OpenMP.cpp

bhandarkar-pranav committed:
clean up a large comment in flang/lib/Lower/OpenMP/OpenMP.cpp

bhandarkar-pranav committed:
Remove an unused variable from genBodyofTargetOp in flang/lib/Lower/OpenMP/OpenMP.cpp

bhandarkar-pranav committed:
Update a comment in MapInfoFinalizationPass and also add one more check that will advance the walk in case the MapInfoOp already has bounds

vzakhari commented on flang/lib/Lower/OpenMP/OpenMP.cpp:
```diff
@@ -268,8 +274,54 @@ static void bindEntryBlockArgs(lower::AbstractConverter &converter,
+            //         (!fir.ref<!fir.char<1,?>>, index)
+
+            // Mapping this to the target will create a use of %arg0 on the
+            // target. Since omp.target is IsolatedFromAbove, %arg0 will have to
```
Sorry, I do not understand why using `%3#2` in the target code results in mapping of `%arg0` - can you please clarify?  It should be just a use of a scalar `index` value, that can be made `firstprivate`.

bhandarkar-pranav replied:
@vzakhari  - `%arg0`  itself isn't a scalar. From my understanding, it'll then have to be mapped so that it can be "unboxed"  (`fir.unbox %arg0`).

The problem really is that `v.len` should have been expressed in terms of `%4:2`. (Please correct me if I am wrong about my understanding of any of these points)

vzakhari replied:
I think I see it now that `cloneBound` clones the defining operation given one of its results.  It was also assuming that the cloned result is `#0` of the operation, which is wrong, and you fixed it for `UnboxCharOp`.

This looks broken to me.  I think what we really want to pass to the OpenMP region is the scalar bound value, and this can be done without cloning the whole defining operation.

The bounds should be scalars, as I understand, so it should be unnecessary to clone the char box to get the bound value.

Consider how this case works:
```
subroutine TestOfCharacter(a0, a1, l)
  integer, intent(in) :: l
  real, intent(in) :: a0(l:)
  real, intent(inout):: a1(l:)

  !$omp target map(to:a0) map(from: a1)
  a1 = a0
  !$omp end target
end subroutine TestOfCharacter
```

The lower bounds are stored into temporaries for passing them via implicit maps to the target region:
```
    %0 = fir.alloca i32
    %1 = fir.alloca i32
    %2 = fir.dummy_scope : !fir.dscope
    %3:2 = hlfir.declare %arg2 dummy_scope %2 {fortran_attrs = #fir.var_attrs<intent_in>, uniq_name = "_QFtestofcharacterEl"} : (!fir.ref<i32>, !fir.dscope) -> (!fir.ref<i32>, !fir.ref<i32>)
    %4 = fir.load %3#0 : !fir.ref<i32>
    fir.store %4 to %0 : !fir.ref<i32>
...
    %9 = fir.load %3#0 : !fir.ref<i32>
    fir.store %9 to %1 : !fir.ref<i32>
...
    %24 = omp.map.info var_ptr(%1 : !fir.ref<i32>, i32) map_clauses(implicit, exit_release_or_enter_alloc) capture(ByCopy) -> !fir.ref<i32> {name = ""}
    %25 = omp.map.info var_ptr(%0 : !fir.ref<i32>, i32) map_clauses(implicit, exit_release_or_enter_alloc) capture(ByCopy) -> !fir.ref<i32> {name = ""}
    omp.target map_entries(%18 -> %arg3, %23 -> %arg4, %24 -> %arg5, %25 -> %arg6 : !fir.ref<!fir.array<?xf32>>, !fir.ref<!fir.array<?xf32>>, !fir.ref<i32>, !fir.ref<i32>) {
      %26 = fir.load %arg6 : !fir.ref<i32>
      %27 = fir.load %arg5 : !fir.ref<i32>
```

I do not know exactly where this happens, but I think the character length should be handled the same way.

Do you know where this array lower bounds handling happens?

bhandarkar-pranav replied:
Since the lower bound `l` is needed in the target OpenMP region in the `hfir.declare` ops for `a0` and `a1`, it is implicitly captured. It is explicitly mapped though because all values used insde the `omp.target` region need to be mapped to incoming block arguments because `omp.target` is `IsolatedFromAbove`. 

The problem in the case that this PR is solving is that this process of creating a temporary for an implicitly captured value is contingent upon such a value having a `definingOp` that is non-null. In the case of block arguments, the definingOp is null.

vzakhari replied:
Isn't the length of the character (i.e. the `index` value itself, not its defining op) also needed in the `hlfir.declare` of the character var?  Or is it different because in the array lower bound there is `!fir.ref<i32>` and in the length case it is `i32`?

> The problem in the case that this PR is solving is that this process of creating a temporary for an implicitly captured value is contingent upon such a value having a definingOp that is non-null. In the case of block arguments, the definingOp is null.

Can you please point me to where this is done?


Again, my general concern here is that we should probably just pass the chracter length value as a literal firstprivate, instead of cloning the defining `fir.unboxchar`.

agozillon replied:
https://github.com/llvm/llvm-project/blob/main/flang/lib/Lower/OpenMP/OpenMP.cpp#L1371 this is the segment I believe, we map or clone across dependencies on the bounds

vzakhari replied:
Thank you! So the temporarization of the direct uses is run after `bindEntryBlockArgs`, and I assume it will kick in if we just generate a direct use of the `index` Value of the character length.  Won't it happen automatically, if you just avoid the cloning, i.e., in the original code, replace this:
```
            converter.bindSymbol(
                sym, fir::CharBoxValue(arg, cloneBound(v.getLen())));
```
with this:
```
            converter.bindSymbol(
                sym, fir::CharBoxValue(arg, v.getLen()));
```
?

bhandarkar-pranav replied:
@vzakhari  - For the reduced testcase that I have added to this PR, not cloning the length works. I'll test this a little more rigorously, while also trying to unearth the motivation for the presence of the clone in the first place (it predates my PR of course)

bhandarkar-pranav committed:
[Flang][mlir] - In fortran lowering, handle block arguments defined outside omp target

This patch handles block arguments that are live-in into the target region of
an omp.target op. This is done by simply reusing the mapping mechanism in place
for values that are not block arguments. Further, in MapInfoFinalizationPass, this
patch adds bounds to maps that map `!fir.ref<!fir.char<k, ?>>` types.
Also, we don't clone bounds when binding entry block arguments any more.

bhandarkar-pranav commented:
@vzakhari  - I have changed my approach and updated this PR. I have tested this locally to the extent that I could and things work fine. Could you please review this PR?

@agozillon @TIFitis @raghavendhra @ergawy @skatrak  - Could you please review this PR? TIA

skatrak commented on flang/lib/Optimizer/OpenMP/MapInfoFinalization.cpp:
```diff
@@ -550,7 +552,39 @@ class MapInfoFinalizationPass
+      // with dynamic length. If so, check if they need bounds added.
+      func->walk([&](mlir::omp::MapInfoOp op) {
+        if (!op.getBounds().empty())
+          return mlir::WalkResult::advance();
```
Nit: Here you can use the `void` callback specialization of `Operation::walk()`, since we don't need to interrupt the walk or skip nested operations.
```suggestion
          return;
```

skatrak commented on flang/include/flang/Optimizer/Builder/DirectivesCommon.h:
```diff
@@ -131,6 +133,31 @@ gatherBoundsOrBoundValues(fir::FirOpBuilder &builder, mlir::Location loc,
+    return builder.create<mlir::omp::MapBoundsOp>(
+        loc, boundTy, /*lower_bound=*/zero,
+        /*upper_bound=*/ub, /*extent=*/extent, /*stride=*/stride,
+        /*stride_in_bytes = */ true, /*start_idx=*/zero);
```
```suggestion
        /*stride_in_bytes=*/ true, /*start_idx=*/zero);
```

skatrak commented on flang/test/Lower/OpenMP/map-character.f90:
```diff
@@ -0,0 +1,46 @@
+
+!CHECK:  %[[A0_BOXCHAR_ALLOCA:.*]] = fir.alloca !fir.boxchar<1>
+!CHECK:  %[[A1_BOXCHAR_ALLOCA:.*]] = fir.alloca !fir.boxchar<1>
+!CHECK:  %[[UNBOXED_ARG0:.*]]:2 = fir.unboxchar %arg0 : (!fir.boxchar<1>) -> (!fir.ref<!fir.char<1,?>>, index)
```
Nit: I'd suggest not hard coding function arguments like `%arg0`, but rather capturing it from a check line (e.g. `CHECK: func.func @...(%[[ARG:.*]]...`) just in case this breaks in the future.

skatrak reviewed (approved):
Thank you Pranav, I gave this a somewhat superficial look and seems reasonable to me. Please wait for @vzakhari's approval before merging, although it's probably best if someone more knowledgeable about OpenMP map handling gave it a look as well.

TIFitis commented on flang/lib/Lower/OpenMP/OpenMP.cpp:
```diff
@@ -217,59 +217,36 @@ static void bindEntryBlockArgs(lower::AbstractConverter &converter,
   assert(args.isValid() && "invalid args");
   fir::FirOpBuilder &firOpBuilder = converter.getFirOpBuilder();
 
-  auto bindSingleMapLike = [&converter,
-                            &firOpBuilder](const semantics::Symbol &sym,
-                                           const mlir::BlockArgument &arg) {
-    // Clones the `bounds` placing them inside the entry block and returns
-    // them.
-    auto cloneBound = [&](mlir::Value bound) {
-      if (mlir::isMemoryEffectFree(bound.getDefiningOp())) {
-        mlir::Operation *clonedOp = firOpBuilder.clone(*bound.getDefiningOp());
-        return clonedOp->getResult(0);
-      }
-      TODO(converter.getCurrentLocation(),
-           "target map-like clause operand unsupported bound type");
-    };
-
-    auto cloneBounds = [cloneBound](llvm::ArrayRef<mlir::Value> bounds) {
-      llvm::SmallVector<mlir::Value> clonedBounds;
-      llvm::transform(bounds, std::back_inserter(clonedBounds),
-                      [&](mlir::Value bound) { return cloneBound(bound); });
-      return clonedBounds;
-    };
-
```
I remember it being essential to clone the bounds rather than using them directly inside the region. But, maybe things have changed which allows this now.

Can you please check if the following snippet works fine with this change?

```
subroutine omp_target_implicit(n)
   integer :: n
   integer :: a(n, 1024)
   !$omp target
      a(2,3) = 10
   !$omp end target
end subroutine omp_target_implicit
```

Also, it would be great if @jeanPerier could +1 this change :)

bhandarkar-pranav replied:
Thanks @TIFitis for taking a look. The above snippet compiled fine with this PR.

jeanPerier replied:
It matters that they are cloned somehow since the region is isolated from above, but I think you may be handling it in the code below on line 1347 that is cloning all the values that do not belong to the region. So, explicit cloning is likely not needed anymore.

```
  mlir::getUsedValuesDefinedAbove(region, valuesDefinedAbove);
  while (!valuesDefinedAbove.empty()) {
```

vzakhari reviewed (approved):
Looks much clearer to me. Thank you!

bhandarkar-pranav committed:
Make changes based on Skatraks review

bhandarkar-pranav committed:
make clang-format happy
